metrics/timescaledb/db.py

```diff
@@ -12,22 +12,42 @@
 log = structlog.get_logger()
 
 
+MAX_PARAMS = 65535  # limit for postgresql
+
+
 def reset_table(table, batch_size=None):
     _drop_table(table, batch_size)
     _ensure_table(table)
     log.info("Reset table", table=table.name)
 
 
 def write(table, rows):
-    max_params = 65535  # limit for postgresql
-    batch_size = max_params // len(table.columns)
+    batch_size = _batch_size(table)
 
     with _get_engine().begin() as connection:
         for values in batched(rows, batch_size):
             connection.execute(insert(table).values(values))
             log.info("Inserted %s rows", len(values), table=table.name)
 
 
+def upsert(table, rows):
+    _ensure_table(table)
+    batch_size = _batch_size(table)
+    non_pk_columns = set(table.columns) - set(table.primary_key.columns)
+
+    with _get_engine().begin() as connection:
+        for values in batched(rows, batch_size):
+            insert_stmt = insert(table).values(values)
+
+            set_ = {c: insert_stmt.excluded[c.name] for c in non_pk_columns}
+
+            insert_stmt = insert_stmt.on_conflict_do_update(
+                constraint=table.primary_key, set_=set_
+            )
```

Comment:
There's a lot of PostgreSQL knowledge packed into these seven lines. Some comments would be useful. Linking to [SQLAlchemy's docs](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#insert-on-conflict-upsert) would be _really_ useful.

SQLAlchemy's `.on_conflict_do_update` method could have better named arguments: `constraint`, for example, is described as the "target" of the conflict. So, as well as/instead of some comments, consider creating variables with more descriptive names.

And as a small aside, why have newlines on 41 and 43? 40, 42, and 44 are closely related conceptually, so I see no reason why they shouldn't be closesly related spatially, too.

Reply:
The argument names make sense to me, but that's probably coming from an RDBMS background. I've added some explanatory comments and a link to the SQLAlchemy docs